File: packages/sdk/src/tests/unit/snapshotMigration0004.test.ts
import { SnapshotSchema } from '@towns-protocol/proto'
import { bin_toHexString } from '@towns-protocol/dlog'
import { create } from '@bufbuild/protobuf'
import { snapshotMigration0004 } from '../../migrations/snapshotMigration0004'
import { streamIdToBytes } from '../../id'
import { makeUniqueSpaceStreamId } from '../testUtils'

describe('snapshotMigration0004', () => {
    test('run migration with common session IDs', () => {
        // Create a test session ID that will appear frequently
        const commonSessionIdBytes = new Uint8Array(
            'common-session-id'.split('').map((c) => c.charCodeAt(0)),
        )
        const commonSessionId = bin_toHexString(commonSessionIdBytes)
        const spaceId = makeUniqueSpaceStreamId()
        const spaceIdBytes = streamIdToBytes(spaceId)

        // Create a snapshot with multiple members having the same session ID
        const snap = create(SnapshotSchema, {
            content: {
                case: 'spaceContent',
                value: {
                    inception: {
                        streamId: spaceIdBytes,
                    },
                },
            },
            members: {
                joined: [
                    {
                        solicitations: [
                            { sessionIds: [commonSessionId, 'unique1_a'] },
                            { sessionIds: [commonSessionId, 'unique1_b'] },
                        ],
                        username: {
                            data: {
                                sessionId: commonSessionId,
                                algorithm: 'test-algo',
                            },
                        },
                        displayName: {
                            data: {
                                sessionIdBytes: commonSessionIdBytes,
                                algorithm: 'test-algo',
                            },
                        },
                    },
                    {
                        solicitations: [{ sessionIds: [commonSessionId, 'unique2'] }],
                        username: {
                            data: {
                                sessionIdBytes: commonSessionIdBytes,
                                algorithm: 'test-algo',
                            },
                        },
                    },
                    {
                        solicitations: [{ sessionIds: [commonSessionId, 'unique3'] }],
                    },
                    {
                        solicitations: [{ sessionIds: ['unique4'] }],
                    },
                ],
            },
        })

        const result = snapshotMigration0004(snap, true)

        // Verify that common session IDs are removed
        for (const member of result.members!.joined) {
            for (const solicitation of member.solicitations) {
                expect(solicitation.sessionIds).not.toContain(commonSessionId)
            }

            // Verify that usernames and display names with common session IDs are cleared
            if (member.username?.data) {
                expect(member.username.data.sessionId).not.toBe(commonSessionId)
                if (member.username.data.sessionIdBytes?.length) {
                    expect(bin_toHexString(member.username.data.sessionIdBytes)).not.toBe(
                        commonSessionId,
                    )
                }
            }

            if (member.displayName?.data) {
                expect(member.displayName.data.sessionId).not.toBe(commonSessionId)
                if (member.displayName.data.sessionIdBytes?.length) {
                    expect(bin_toHexString(member.displayName.data.sessionIdBytes)).not.toBe(
                        commonSessionId,
                    )
                }
            }
        }
    })

    test('run migration with empty members', () => {
        const snap = create(SnapshotSchema, {})
        const result = snapshotMigration0004(snap, true)
        expect(result).toEqual(snap)
    })

    test('run migration with no common session IDs', () => {
        const snap = create(SnapshotSchema, {
            members: {
                joined: [
                    {
                        solicitations: [{ sessionIds: ['unique1', 'unique2'] }],
                        username: {
                            data: {
                                sessionId: 'unique3',
                                algorithm: 'test-algo',
                            },
                        },
                    },
                    {
                        solicitations: [{ sessionIds: ['unique4', 'unique5'] }],
                    },
                ],
            },
        })

        const result = snapshotMigration0004(snap, true)
        // Verify the snapshot remains unchanged when no common session IDs are found
        expect(result).toEqual(snap)
    })
})

File: core/node/events/migrations/snapshot_migration_0004_test.go
package migrations

import (
	"encoding/hex"
	"testing"

	"github.com/stretchr/testify/require"

	"google.golang.org/protobuf/proto"

	. "github.com/towns-protocol/towns/core/node/protocol"
	"github.com/towns-protocol/towns/core/node/shared"
)

func TestSnapshotMigration0004(t *testing.T) {
	// Test case 1: Run migration with common session IDs
	t.Run("with common session IDs", func(t *testing.T) {
		commonSessionIdBytes := []byte("common-session-id")
		commonSessionId := hex.EncodeToString(commonSessionIdBytes)
		spaceId, err := shared.MakeSpaceId()
		require.NoError(t, err)

		snap := &Snapshot{
			Content: &Snapshot_SpaceContent{
				SpaceContent: &SpacePayload_Snapshot{
					Inception: &SpacePayload_Inception{
						StreamId: spaceId[:],
					},
				},
			},
			Members: &MemberPayload_Snapshot{
				Joined: []*MemberPayload_Snapshot_Member{
					{
						Solicitations: []*MemberPayload_KeySolicitation{
							{SessionIds: []string{commonSessionId, "unique1_a"}},
							{SessionIds: []string{commonSessionId, "unique1_b"}},
						},
						Username: &WrappedEncryptedData{
							Data: &EncryptedData{
								SessionId: commonSessionId,
								Algorithm: "test-algo",
							},
						},
						DisplayName: &WrappedEncryptedData{
							Data: &EncryptedData{
								SessionIdBytes: commonSessionIdBytes,
								Algorithm:      "test-algo",
							},
						},
					},
					{
						Solicitations: []*MemberPayload_KeySolicitation{
							{SessionIds: []string{commonSessionId, "unique2"}},
						},
						Username: &WrappedEncryptedData{
							Data: &EncryptedData{
								SessionIdBytes: commonSessionIdBytes,
								Algorithm:      "test-algo",
							},
						},
					},
					{
						Solicitations: []*MemberPayload_KeySolicitation{
							{SessionIds: []string{commonSessionId, "unique3"}},
						},
					},
					{
						Solicitations: []*MemberPayload_KeySolicitation{
							{SessionIds: []string{"unique4"}},
						},
					},
				},
			},
		}

		snapshot_migration_0004_(snap, true)

		// Verify that common session IDs are removed
		for _, member := range snap.GetMembers().GetJoined() {
			for _, solicitation := range member.GetSolicitations() {
				require.NotContains(t, solicitation.GetSessionIds(), commonSessionId)
			}

			// Verify that usernames and display names with common session IDs are cleared
			if member.GetUsername() != nil {
				require.NotEqual(t, member.GetUsername().GetData().GetSessionId(), commonSessionId)
				if len(member.GetUsername().GetData().GetSessionIdBytes()) > 0 {
					require.NotEqual(t, string(member.GetUsername().GetData().GetSessionIdBytes()), commonSessionId)
				}
			}

			if member.GetDisplayName() != nil {
				require.NotEqual(t, member.GetDisplayName().GetData().GetSessionId(), commonSessionId)
				if len(member.GetDisplayName().GetData().GetSessionIdBytes()) > 0 {
					require.NotEqual(t, string(member.GetDisplayName().GetData().GetSessionIdBytes()), commonSessionId)
				}
			}
		}
	})

	// Test case 2: Run migration with empty members
	t.Run("with empty members", func(t *testing.T) {
		snap := &Snapshot{}
		snapshot_migration_0004_(snap, true)
		// The snapshot should remain the same reference
		require.NotNil(t, snap)
	})

	// Test case 3: Run migration with no common session IDs
	t.Run("with no common session IDs", func(t *testing.T) {
		snap := &Snapshot{
			Members: &MemberPayload_Snapshot{
				Joined: []*MemberPayload_Snapshot_Member{
					{
						UserAddress: []byte("member1"),
						Solicitations: []*MemberPayload_KeySolicitation{
							{SessionIds: []string{"unique1", "unique2"}},
						},
						Username: &WrappedEncryptedData{
							Data: &EncryptedData{
								SessionId: "unique3",
								Algorithm: "test-algo",
							},
						},
					},
					{
						UserAddress: []byte("member2"),
						Solicitations: []*MemberPayload_KeySolicitation{
							{SessionIds: []string{"unique4", "unique5"}},
						},
					},
				},
			},
		}

		snapCopy, err := proto.Marshal(snap)
		require.NoError(t, err)
		snapshot_migration_0004_(snap, true)
		// Verify the snapshot remains unchanged when no common session IDs are found
		resultBytes, err := proto.Marshal(snap)
		require.NoError(t, err)
		require.Equal(t, snapCopy, resultBytes)
	})
}
